## 3. Processamento/Tratamento de Dados

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# Mostrar todas as colunas dos dataframes
pd.set_option('display.max_columns', None)

### Conhecendo a base:

In [2]:
# IMPORTAÇÃO DA BASE DE EMPRESAS
empresas_df = pd.read_csv(r'C:\DATASETS\Empresas Rotuladas\empresas_rotuladas-consolidado.csv', dtype=str)
empresas_df

cnpj matriz_filial  \
0        00000942000170             1   
1        00001102000121             1   
2        00005201000181             1   
3        00005201000262             2   
4        00010501000159             1   
...                 ...           ...   
1559938  98593668000194             1   
1559939  98669997000252             2   
1559940  98670235000195             1   
1559941  98670656000116             1   
1559942  98674195000150             1   

                                              razao_social  \
0                                      NILZA LAVISO ARAUJO   
1                          JOSE GERALDO PEREIRA SANTANA ME   
2        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO LTDA   
3        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO LTDA   
4        B S PETRO COM E DIST DE DERIVADOS DE PETROLEO ...   
...                                                    ...   
1559938                    AUTO VIACAO VENANCIO AIRES LTDA   
1559939          E. R. AMANTINO INDUSTRIA METALURGICA LTDA   
1559940                    MAZZAROLLO - AGRO MAQUINAS LTDA   
1559941                            RESTAURANTE FARINA LTDA   
1559942                           AEROCLUBE DE VERANOPOLIS   

                                        nome_fantasia situacao data_situacao  \
0                                        MIL CARIMBOS       08      20180201   
1                                                 NaN       08      20180201   
2                                                 NaN       08      20180326   
3        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO       08      20180326   
4                                           B S PETRO       08      20180326   
...                                               ...      ...           ...   
1559938                                        VIASUL       02      20040515   
1559939                                     MICROVERA       02      20051103   
1559940                                           NaN       02      20020720   
1559941                                           NaN       02      20050416   
1559942                                           NaN       02      20050924   

        motivo_situacao cod_nat_juridica data_inicio_ativ cnae_fiscal  \
0                    67             2135         19940505     3299002   
1                    67             2135         19940505     4729699   
2                    15             2062         19940517     4681802   
3                    15             2062         20011211     4681802   
4                    15             2062         19940527     4681801   
...                 ...              ...              ...         ...   
1559938              00             2062         19680327     4922102   
1559939              00             2062         19820303     2451200   
1559940              00             2062         19660824     4789099   
1559941              00             2062         19660823     5611201   
1559942              00             3999         19671024     8599601   

        tipo_logradouro                logradouro numero  complemento  \
0                   RUA       PREFEITO S A T PIZA    235          NaN   
1                   RUA              ANGELO GUIDI    184          NaN   
2                   RUA              DOS ANDRADAS    421  CONJUNTO 52   
3               AVENIDA  DEPUTADO NARCISO PIERONI    450          NaN   
4               AVENIDA              MONS SECKLER   1494          NaN   
...                 ...                       ...    ...          ...   
1559938             RUA     VOLUNTARIOS DA PATRIA   1889          NaN   
1559939             RUA            GETULIO VARGAS    642          NaN   
1559940         AVENIDA            OSVALDO ARANHA    452          NaN   
1559941         ESTRADA           ESTRADA RST 470    691          NaN   
1559942         ALAMEDA             SANTOS DUMONT    S N          NaN   

                 bairro       cep  uf            municipio ddd_1 t

In [3]:
empresas_df.describe()

cnpj matriz_filial                      razao_social  \
count          1559943       1559943                           1559943   
unique         1559255             2                           1547558   
top     12486624000140             1  COAMO AGROINDUSTRIAL COOPERATIVA   
freq                 2       1547369                               172   

           nome_fantasia situacao data_situacao motivo_situacao  \
count             684671  1559943       1559943         1559943   
unique            595304        5          6866              31   
top     ESPACO DA BELEZA       08      20180201              67   
freq                 168  1385738       1372085         1372057   

       cod_nat_juridica data_inicio_ativ cnae_fiscal tipo_logradouro  \
count           1559943          1559943     1559943         1557957   
unique               65            14825        1232             318   
top                2135         20101022     4781400             RUA   
freq            1430198             3377      151887         1107051   

       logradouro   numero complemento   bairro       cep       uf  municipio  \
count     1559939  1559851      638343  1559465   1559917  1559943    1559943   
unique     417052    39418      140689    92181    370833       28       5280   
top        BRASIL       SN        CASA   CENTRO  42700130       SP  SAO PAULO   
freq         4058    54320      117778   175275      2575   411029     143300   

          ddd_1 telefone_1         email qualif_resp capital_social    porte  \
count   1559943    1539066        948320     1559943        1559943  1559943   
unique      276    1381158        857767          27          14396        4   
top          11          0  nt@nt.com.br          50            1.0       01   
freq     237304       4820           282     1430143         806974  1493736   

       opc_simples  opc_mei inexistente anos_de_atividade_empresa  \
count      1559943  1559943     1559943                   1559943   
unique           3        2           2                        93   
top              6    False        True                         5   
freq       1399627  1533044     1375877                    296196   

       telefone_provavelmente_invalido Atendimentos_Procon  \
count                          1559943             1559943   
unique                               2                 456   
top                              False                   0   
freq                           1551495             1551151   

       empresa_devedora_uniao empresa_venceu_licitacao  \
count                 1559943                  1559943   
unique                      2                        2   
top                         0                        0   
freq                  1521397                  1550716   

       empresa_participou_licitacao empresa_recurso_recebido_governo_federal  \
count                       1559943                                  1559943   
unique                            2                                    14614   
top                               0                                      0.0   
freq                        1548679                                  1544254   

       empresa_sancionada qte_socios tempo_de_sociedade  \
count             1559943    1507238            1507238   
unique                  2         77                 73   
top                     0          1                  5   
freq              1559657    1411583             298979   

       auxilio_emergencial_bloqueado socio_solicitou_auxilio_emergencial  \
count                        1507238                             1507238   
unique                             2                                   2   
top                                0                                   1   
freq                         1467942                              798201   

       socio_solicitou_bolsa_familia socio_candidato socio_devedor_uniao  \
count                        1507238     

In [4]:
empresas_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1559943 entries, 0 to 1559942
Data columns (total 48 columns):
cnpj                                        1559943 non-null object
matriz_filial                               1559943 non-null object
razao_social                                1559943 non-null object
nome_fantasia                               684671 non-null object
situacao                                    1559943 non-null object
data_situacao                               1559943 non-null object
motivo_situacao                             1559943 non-null object
cod_nat_juridica                            1559943 non-null object
data_inicio_ativ                            1559943 non-null object
cnae_fiscal                                 1559943 non-null object
tipo_logradouro                             1557957 non-null object
logradouro                                  1559939 non-null object
numero                                      1559851 non-null object
comp

#### Identificando percentual de valores ausentes por coluna:

In [5]:
(empresas_df.isnull().sum() / empresas_df.shape[0] * 100).sort_values(ascending=False)

complemento                                 59.079082
nome_fantasia                               56.109230
email                                       39.208035
socio_solicitou_auxilio_emergencial          3.378649
socio_solicitou_seguro_defeso                3.378649
qte_socios                                   3.378649
tempo_de_sociedade                           3.378649
auxilio_emergencial_bloqueado                3.378649
socio_sancionado                             3.378649
socio_solicitou_bolsa_familia                3.378649
socio_candidato                              3.378649
socio_devedor_uniao                          3.378649
empresas_inexistentes_socio_participa        3.378649
socio_recurso_recebido_governo_federal       3.378649
socio_pessoa_politicamente_exposta           3.378649
qte_cnpj_socio_participa                     3.378649
telefone_1                                   1.338318
tipo_logradouro                              0.127312
bairro                      

### Colunas não necessárias e/ou com alto índice de nulos:

Analisando os dados acima, identificamos colunas com alto índice de nulos (>50%), que poderão ser removidas. Também foram analisados os dicionários de dados das bases envolvidas (especialmente da base nacional de empresas, extraída do site da RFB), e foram identificadas algumas colunas que não são relevantes para essa análise.

* alto índice de nulos: complemento, nome_fantasia
* não relevantes: data_situacao, tipo_logradouro, logradouro, numero, complemento, bairro, cep, ddd1_, telefone_1, email
* data_inicio_ativ: já criamos uma nova coluna 'anos_de_atividade_empresa' para substituir a 'data_inicio_ativ'

Podemos remover algumas delas:

In [6]:
empresas_df = empresas_df.drop(columns=['nome_fantasia', 'data_inicio_ativ', 'data_situacao'])

Não vamos remover agora as colunas de endereços, email e telefone. Vamos primeiro fazer feature engeneering sobre elas, e depois removê-las (conforme o caso).

### Feature engineering:

##### Coluna 'email':

O e-mail por si só não é interessante para essa análise. Mas há duas informações relevantes que podemos extrair da coluna de e-mail: 1) se o e-mail foi ou não preenchido; 2) e qual é o domínio do e-mail. Um domínio específico (como bb.com.br, santanense.com.br) indica uma empresa mais estruturada por trás (mesmo que seja o e-mail da contabilidade). Já um domínio genérico (gmail.com, hotmail.com) não é tão forte nesse sentido. <br>Então, vamos extrair os domínios dos emails, e preencher com NI para os nulos. Vamos considerar as duas informações em uma mesma coluna (vamos criar a coluna 'email_processado').

In [7]:
# Correção de registros como '@@hotmail.com':
empresas_df["email"] = empresas_df["email"].str.replace('@@', '@', regex=False)
# Criação da coluna email_processado>
empresas_df["email_processado"] = np.where((empresas_df["email"].astype(str).str.contains("@"))&(~empresas_df["email"].astype(str).str.endswith("@")), empresas_df["email"].astype(str).str.split("@").str[1], np.where(empresas_df["email"].isnull(), 'NI', empresas_df["email"]))

In [8]:
empresas_df[["email","email_processado"]].describe

<bound method NDFrame.describe of                                    email   email_processado
0                                    NaN                 NI
1                                    NaN                 NI
2             petroalfa@petroalfa.cnt.br   petroalfa.cnt.br
3             PETROALFA@PETROALFA.COM.BR   PETROALFA.COM.BR
4                                    NaN                 NI
...                                  ...                ...
1559938          VIASUL@VIASULBUS.COM.BR   VIASULBUS.COM.BR
1559939  CONTABILIDADE@ERAMANTINO.COM.BR  ERAMANTINO.COM.BR
1559940                              NaN                 NI
1559941                              NaN                 NI
1559942                              NaN                 NI

[1559943 rows x 2 columns]>

In [10]:
# Vamos ver o resultado:
empresas_df["email_processado"] = empresas_df["email_processado"].str.upper()
empresas_df["email_processado"].value_counts()

NI                             611623
HOTMAIL.COM                    447041
GMAIL.COM                      188999
YAHOO.COM.BR                    71222
BOL.COM.BR                      29532
                                ...  
ASATNET.COM.BR                      1
AROXA.COM.BR                        1
CONCEITOASSESSORIASA.COM.BR         1
LIBERTYTURISMO.COM                  1
BRAVIETOUR.COM.BR                   1
Name: email_processado, Length: 53000, dtype: int64

Podemos remover a coluna de email original:

In [11]:
empresas_df = empresas_df.drop(columns=['email'])

##### Colunas ddd_1 e telefone_1:

Por si só, não são relevantes, mas vamos extrair delas duas informações: 1) se o telefone foi preenchido ou não; 2) se o telefone é utilizado por outra empresa.

Primeiro, vamos ver se o telefone é utilizado por mais de uma empresa:

In [12]:
# Concatenar ddd + telefone:
empresas_df['telefone_processado'] = empresas_df['ddd_1'] + empresas_df['telefone_1']
empresas_df['telefone_processado'].describe

<bound method NDFrame.describe of 0                 NaN
1                 NaN
2          1181651700
3          1133380127
4                 NaN
              ...    
1559938    5137931700
1559939    5434419100
1559940           NaN
1559941           NaN
1559942           NaN
Name: telefone_processado, Length: 1559943, dtype: object>

In [13]:
# Calcular quantas empresas compartilham do mesmo telefone:
telefones_df = empresas_df[['cnpj','telefone_processado']]
    .groupby(['telefone_processado'])['cnpj'].count().reset_index()
telefones_df

telefone_processado  cnpj
0                **********     1
1                        00  3437
2                0000000000   484
3              000000000000     8
4                0000000001     1
...                     ...   ...
1421989          9999998024     1
1421990          9999999999    57
1421991          NI00773261     1
1421992          NI34731270     1
1421993              NI3628     1

[1421994 rows x 2 columns]

In [14]:
# Conferindo:
empresas_df[empresas_df['telefone_processado'] == '000000000000']

cnpj matriz_filial  \
1745     03639040000176             1   
1389454  02871638000123             1   
1391627  03266986000134             1   
1392251  03373470000199             1   
1393584  03575393000150             1   
1404703  05330553000126             1   
1413744  07212650000140             1   
1550377  72452188000110             1   

                                      razao_social situacao motivo_situacao  \
1745                    BONZANINI & BONZANINI LTDA       08              15   
1389454                  TRANSPORTES LUCHETTA LTDA       02              00   
1391627                   VERMELAV LAVANDERIA LTDA       04              63   
1392251                  BOCA DA ONCA ECOTOUR LTDA       02              00   
1393584                A A SAUERESSIG FERRARI LTDA       02              00   
1404703                         AUREO LUIZ RAVAGIO       02              00   
1413744                    LIDIANE VIEIRA DA SILVA       04              63   
1550377  TRACAM TRATOR CAMPOS MECANICA DIESEL LTDA       02              00   

        cod_nat_juridica cnae_fiscal tipo_logradouro  \
1745                2062     4681801         AVENIDA   
1389454             2062     4924800             NaN   
1391627             2062     9601701         RODOVIA   
1392251             2062     9329899         RODOVIA   
1393584             2062     5510801         AVENIDA   
1404703             2135     4744001             RUA   
1413744             2135     4781400         AVENIDA   
1550377             2062     3314711             NaN   

                       logradouro  numero complemento              bairro  \
1745      ISMAEL J. DO NASCIMENTO  1357-W        SALA    JARDIM TANGARA I   
1389454  MARGENS DA BR 470 KM 403     S/N         NaN   REASSENT.GERASSUL   
1391627          PRESIDENTE DUTRA    4648           B  SAO JOAO DE MERITI   
1392251     BODOQUENA PARA BONITO     S/N       KM 26          ZONA RURAL   
1393584          MATE  LARANJEIRA   1.155      TERREO              CENTRO   
1404703              DALTRO FILHO     108        SALA              CENTRO   
1413744           MANOEL MURTINHO    1322         NaN              CENTRO   
1550377  MARGENS DA BR 282 KM 343     S/N         NaN              TABOAO   

              cep  uf           municipio ddd_1 telefone_1 qualif_resp  \
1745     78300000  MT    TANGARA DA SERRA  0000   00000000          49   
1389454  89620000  SC        CAMPOS NOVOS  0000   00000000          49   
1391627  25586140  RJ  SAO JOAO DE MERITI  0000   00000000          49   
1392251  79390000  MS           BODOQUENA  0000   00000000          49   
1393584  85980000  PR              GUAIRA  0000   00000000          49   
1404703  98470000  RS            PLANALTO  0000   00000000          50   
1413744  79210000  MS           ANASTACIO  0000   00000000          50   
1550377  89620000  SC        CAMPOS NOVOS  0000   00000000          49   

        capital_social porte opc_simples opc_mei inexistente  \
1745               0.0    01           0   False        True   
1389454       183000.0    01           5   False       False   
1391627            0.0    01           6   False       False   
1392251        60000.0    01           5   False       False   
1393584        59000.0    01           5   False       False   
1404703        20000.0    01           5   False       False   
1413744            0.0    01           0   False       False   
1550377        40000.0    01           5   False       False   

        anos_de_atividade_empresa telefone_provavelmente_invalido  \
1745                           19                            True   
1389454                        23                            True   
1391627                        22                            True   
1392251                        22                            True   
1393584                        22                            True   
1404703                        19                            True   
1413744    

Conforme esperado. São empresas que têm o mesmo número de telefone (independente se está correto ou não, pois isso será visto na coluna de telefone_possivelmente_invalido).

Vamos fazer o join de telefones_df e empresas_df:

In [15]:
# Renomear a coluna:
telefones_df = telefones_df.rename(columns={'cnpj': 'qte_empresas_com_mesmo_telefone'})

In [16]:
empresas_df = empresas_df.merge(telefones_df, on='telefone_processado', how='left')
empresas_df

cnpj matriz_filial  \
0        00000942000170             1   
1        00001102000121             1   
2        00005201000181             1   
3        00005201000262             2   
4        00010501000159             1   
...                 ...           ...   
1559938  98593668000194             1   
1559939  98669997000252             2   
1559940  98670235000195             1   
1559941  98670656000116             1   
1559942  98674195000150             1   

                                              razao_social situacao  \
0                                      NILZA LAVISO ARAUJO       08   
1                          JOSE GERALDO PEREIRA SANTANA ME       08   
2        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO LTDA       08   
3        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO LTDA       08   
4        B S PETRO COM E DIST DE DERIVADOS DE PETROLEO ...       08   
...                                                    ...      ...   
1559938                    AUTO VIACAO VENANCIO AIRES LTDA       02   
1559939          E. R. AMANTINO INDUSTRIA METALURGICA LTDA       02   
1559940                    MAZZAROLLO - AGRO MAQUINAS LTDA       02   
1559941                            RESTAURANTE FARINA LTDA       02   
1559942                           AEROCLUBE DE VERANOPOLIS       02   

        motivo_situacao cod_nat_juridica cnae_fiscal tipo_logradouro  \
0                    67             2135     3299002             RUA   
1                    67             2135     4729699             RUA   
2                    15             2062     4681802             RUA   
3                    15             2062     4681802         AVENIDA   
4                    15             2062     4681801         AVENIDA   
...                 ...              ...         ...             ...   
1559938              00             2062     4922102             RUA   
1559939              00             2062     2451200             RUA   
1559940              00             2062     4789099         AVENIDA   
1559941              00             2062     5611201         ESTRADA   
1559942              00             3999     8599601         ALAMEDA   

                       logradouro numero  complemento          bairro  \
0             PREFEITO S A T PIZA    235          NaN  JARDIM PARAIZO   
1                    ANGELO GUIDI    184          NaN          CENTRO   
2                    DOS ANDRADAS    421  CONJUNTO 52  SANTA IFIGENIA   
3        DEPUTADO NARCISO PIERONI    450          NaN       MOSTARDAS   
4                    MONS SECKLER   1494          NaN      VL AMERICA   
...                           ...    ...          ...             ...   
1559938     VOLUNTARIOS DA PATRIA   1889          NaN          CENTRO   
1559939            GETULIO VARGAS    642          NaN       RENOVACAO   
1559940            OSVALDO ARANHA    452          NaN          CENTRO   
1559941           ESTRADA RST 470    691          NaN          CENTRO   
1559942             SANTOS DUMONT    S N          NaN          CENTRO   

              cep  uf            municipio ddd_1 telefone_1 qualif_resp  \
0        16600000  SP              PIRAJUI    NI        NaN          50   
1        14180000  SP               PONTAL    NI        NaN          50   
2        01208000  SP            SAO PAULO    11   81651700          49   
3        13820000  SP  MONTE ALEGRE DO SUL    11   33380127          49   
4        18540000  SP          PORTO FELIZ    NI        NaN          49   
...           ...  ..                  ...   ...        ...         ...   
1559938  95800000  RS       VENANCIO AIRES    51   37931700          49   
1559939  95330000  RS          VERANOPOLIS    54   34419100          49   
1559940  95330000  RS          VERANOPOLIS    NI        NaN          49   
1559941  95330000  RS          VERANOPOLIS    NI        NaN          49   
1559942  95330000  RS          VERANOPOLIS    NI        NaN          16   

        capital_social porte o

In [17]:
# Conferindo:
empresas_df[empresas_df['telefone_processado'] == '000000000000']

cnpj matriz_filial  \
1745     03639040000176             1   
1389454  02871638000123             1   
1391627  03266986000134             1   
1392251  03373470000199             1   
1393584  03575393000150             1   
1404703  05330553000126             1   
1413744  07212650000140             1   
1550377  72452188000110             1   

                                      razao_social situacao motivo_situacao  \
1745                    BONZANINI & BONZANINI LTDA       08              15   
1389454                  TRANSPORTES LUCHETTA LTDA       02              00   
1391627                   VERMELAV LAVANDERIA LTDA       04              63   
1392251                  BOCA DA ONCA ECOTOUR LTDA       02              00   
1393584                A A SAUERESSIG FERRARI LTDA       02              00   
1404703                         AUREO LUIZ RAVAGIO       02              00   
1413744                    LIDIANE VIEIRA DA SILVA       04              63   
1550377  TRACAM TRATOR CAMPOS MECANICA DIESEL LTDA       02              00   

        cod_nat_juridica cnae_fiscal tipo_logradouro  \
1745                2062     4681801         AVENIDA   
1389454             2062     4924800             NaN   
1391627             2062     9601701         RODOVIA   
1392251             2062     9329899         RODOVIA   
1393584             2062     5510801         AVENIDA   
1404703             2135     4744001             RUA   
1413744             2135     4781400         AVENIDA   
1550377             2062     3314711             NaN   

                       logradouro  numero complemento              bairro  \
1745      ISMAEL J. DO NASCIMENTO  1357-W        SALA    JARDIM TANGARA I   
1389454  MARGENS DA BR 470 KM 403     S/N         NaN   REASSENT.GERASSUL   
1391627          PRESIDENTE DUTRA    4648           B  SAO JOAO DE MERITI   
1392251     BODOQUENA PARA BONITO     S/N       KM 26          ZONA RURAL   
1393584          MATE  LARANJEIRA   1.155      TERREO              CENTRO   
1404703              DALTRO FILHO     108        SALA              CENTRO   
1413744           MANOEL MURTINHO    1322         NaN              CENTRO   
1550377  MARGENS DA BR 282 KM 343     S/N         NaN              TABOAO   

              cep  uf           municipio ddd_1 telefone_1 qualif_resp  \
1745     78300000  MT    TANGARA DA SERRA  0000   00000000          49   
1389454  89620000  SC        CAMPOS NOVOS  0000   00000000          49   
1391627  25586140  RJ  SAO JOAO DE MERITI  0000   00000000          49   
1392251  79390000  MS           BODOQUENA  0000   00000000          49   
1393584  85980000  PR              GUAIRA  0000   00000000          49   
1404703  98470000  RS            PLANALTO  0000   00000000          50   
1413744  79210000  MS           ANASTACIO  0000   00000000          50   
1550377  89620000  SC        CAMPOS NOVOS  0000   00000000          49   

        capital_social porte opc_simples opc_mei inexistente  \
1745               0.0    01           0   False        True   
1389454       183000.0    01           5   False       False   
1391627            0.0    01           6   False       False   
1392251        60000.0    01           5   False       False   
1393584        59000.0    01           5   False       False   
1404703        20000.0    01           5   False       False   
1413744            0.0    01           0   False       False   
1550377        40000.0    01           5   False       False   

        anos_de_atividade_empresa telefone_provavelmente_invalido  \
1745                           19                            True   
1389454                        23                            True   
1391627                        22                            True   
1392251                        22                            True   
1393584                        22                            True   
1404703                        19                            True   
1413744    

Conforme esperado. 8 empresas na coluna 'qte_empresas_com_mesmo_telefone'.<br>Vamos colocar 0 nos nulos:

In [18]:
empresas_df.qte_empresas_com_mesmo_telefone.describe()

count    1.539066e+06
mean     9.777599e+00
std      1.626830e+02
min      1.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      1.000000e+00
max      3.437000e+03
Name: qte_empresas_com_mesmo_telefone, dtype: float64

In [19]:
empresas_df = empresas_df.fillna(value={'qte_empresas_com_mesmo_telefone': 0})
empresas_df.qte_empresas_com_mesmo_telefone = empresas_df.qte_empresas_com_mesmo_telefone.astype(int)
empresas_df.qte_empresas_com_mesmo_telefone.describe

<bound method NDFrame.describe of 0          0
1          0
2          1
3          1
4          0
          ..
1559938    2
1559939    1
1559940    0
1559941    0
1559942    0
Name: qte_empresas_com_mesmo_telefone, Length: 1559943, dtype: int32>

In [20]:
empresas_df.qte_empresas_com_mesmo_telefone.value_counts()

1     1365309
2       81446
3       21411
0       20877
4       12124
       ...   
70         70
65         65
55         55
54         54
47         47
Name: qte_empresas_com_mesmo_telefone, Length: 86, dtype: int64

Obtivemos a primeira informação. Agora vamos obter a segunda: se o telefone foi ou não preenchido:

In [21]:
empresas_df['telefone_preenchido'] = ~empresas_df['telefone_1'].isnull()
print(empresas_df['telefone_preenchido'].value_counts())
print(empresas_df['telefone_1'].isnull().sum())

True     1539066
False      20877
Name: telefone_preenchido, dtype: int64
20877


Tudo certo. Vamos remover as colunas não mais necessárias:

In [22]:
empresas_df = empresas_df.drop(columns=['ddd_1', 'telefone_1', 'telefone_processado'])

In [23]:
# Pequena correção
empresas_df.telefone_preenchido = np.where(empresas_df.telefone_preenchido==True, 1, 0)

##### Colunas de endereços:

As colunas de endereço por si só não são interessantes para essa análise. Mas vamos extrair delas uma informação importante: quantas empresas compartilham do exato mesmo endereço? (pode acontecer, mas é um indicador de suspeita, prática comum em empresas fictícias ou montadas para algum propósito irregular, como em fraudes de licitações).

In [24]:
# Colunas de endereço:
cols = ['tipo_logradouro', 'logradouro', 'numero', 'complemento', 'bairro', 'municipio', 'uf']

# Primeiro, fazemos preparação das colunas de enderaço para o join e posterior merge:
empresas_df = empresas_df.fillna(value={'tipo_logradouro':'', 'logradouro':'', 'numero':'', 'complemento':'', 'bairro':'', 'municipio':'', 'uf':''})
for col in cols:
    empresas_df[col] = empresas_df[col].str.upper()
    
# Concatenar as colunas de endereço:
empresas_df['endereco_processado'] = empresas_df[cols].apply(lambda row: '_'.join(row.values.astype(str)), axis=1)
empresas_df['endereco_processado'].describe

<bound method NDFrame.describe of 0          RUA_PREFEITO S A T PIZA_235__JARDIM PARAIZO_PI...
1                     RUA_ANGELO GUIDI_184__CENTRO_PONTAL_SP
2          RUA_DOS ANDRADAS_421_CONJUNTO 52_SANTA IFIGENI...
3          AVENIDA_DEPUTADO NARCISO PIERONI_450__MOSTARDA...
4          AVENIDA_MONS SECKLER_1494__VL AMERICA_PORTO FE...
                                 ...                        
1559938    RUA_VOLUNTARIOS DA PATRIA_1889__CENTRO_VENANCI...
1559939     RUA_GETULIO VARGAS_642__RENOVACAO_VERANOPOLIS_RS
1559940    AVENIDA_OSVALDO ARANHA_452__CENTRO_VERANOPOLIS_RS
1559941    ESTRADA_ESTRADA RST 470_691__CENTRO_VERANOPOLI...
1559942     ALAMEDA_SANTOS DUMONT_S N__CENTRO_VERANOPOLIS_RS
Name: endereco_processado, Length: 1559943, dtype: object>

In [25]:
# Calcular quantas empresas compartilham do mesmo endereço:
enderecos_df = empresas_df[['cnpj','endereco_processado']].groupby(['endereco_processado'])['cnpj'].count().reset_index()
enderecos_df = enderecos_df.rename(columns={'cnpj': 'qte_empresas_com_mesmo_endereco'})
enderecos_df

endereco_processado  \
0             10A AVENIDA_13 DE MAIO_834__CENTRO_PANCAS_ES   
1        10A AVENIDA_7 DE SETEMBRO_1104_SALA 01_CENTRO_...   
2        10A AVENIDA_7 DE SETEMBRO_850__CENTRO_NOVO ACO...   
3          10A AVENIDA_9 DE AGOSTO_1959__CENTRO_JAGUARE_ES   
4        10A AVENIDA_ALFREDO DANTE FASSINI_S/N_LOTE 13 ...   
...                                                    ...   
1503750     ÁREA_VILA SETEL_1__UFV CAMPUS VICOSA_VICOSA_MG   
1503751  ÁREA_ZONA LITORANEA_SN_GALPAO_PEDRA MOLE_IGRAP...   
1503752  ÁREA_ZONA RURAL - SAO JOAO_9999__SAO JOAO_MARI...   
1503753     ÁREA_ZONA RURAL_12__PIMENTEIRA_SAO BENEDITO_CE   
1503754  ÁREA_ÁREA ESPECIAL Nº 07 SETOR CSC TAGUATINGA ...   

         qte_empresas_com_mesmo_endereco  
0                                      1  
1                                      1  
2                                      1  
3                                      1  
4                                      1  
...                                  ...  
1503750                                1  
1503751                                1  
1503752                                1  
1503753                                1  
1503754                                1  

[1503755 rows x 2 columns]

In [26]:
# Conferindo:
enderecos_df['qte_empresas_com_mesmo_endereco'].value_counts()

1      1457333
2        41182
3         3734
4          762
5          288
6          149
7           95
8           49
9           37
10          26
11          21
12          15
16           6
15           6
13           5
14           5
20           5
30           4
21           4
24           3
19           3
18           3
26           3
22           2
23           1
25           1
17           1
27           1
80           1
31           1
33           1
36           1
39           1
40           1
44           1
48           1
51           1
65           1
112          1
Name: qte_empresas_com_mesmo_endereco, dtype: int64

In [27]:
enderecos_df[enderecos_df['qte_empresas_com_mesmo_endereco'] > 50]

endereco_processado  \
25042    AVENIDA_ADOLFO PINHEIRO_170__SANTO AMARO_SAO P...   
76357    AVENIDA_COMENDADOR FEIZ ZARZUR_843_SALA 01_JAR...   
112769   AVENIDA_ENGENHEIRO SOUZA FILHO_SN__ITANHANGA_R...   
1302316  RUA_SAINT ROMAN_200__COPACABANA_RIO DE JANEIRO_RJ   

         qte_empresas_com_mesmo_endereco  
25042                                112  
76357                                 80  
112769                                51  
1302316                               65

In [28]:
empresas_df[empresas_df['endereco_processado']=='RUA_SAINT ROMAN_200__COPACABANA_RIO DE JANEIRO_RJ']

cnpj matriz_filial  \
21503    11470156000152             1   
60873    11821417000131             1   
71846    11908988000108             1   
88541    12030632000187             1   
106682   12182277000161             1   
...                 ...           ...   
1286786  21053847000104             1   
1304084  21196378000182             1   
1340299  21502367000183             1   
1341985  21517812000189             1   
1349247  21588278000100             1   

                                              razao_social situacao  \
21503                     ROBERTO VICENTE COSTA DOS SANTOS       08   
60873                 VANESSA RENEE ALVES REIS 44428480253       08   
71846              NIVALDO CAVALCANTE DE MOURA 98736906700       08   
88541                   MAGNO TIAGO DOS SANTOS 01123741743       08   
106682                 ROBSON BRANDAO CASCIANO 07641572723       08   
...                                                    ...      ...   
1286786               LUCIANA LOPES DE ALMEIDA 01490391746       08   
1304084  JOAQUIM GUILHERME BASTOS MARINHO VIANA 1462450...       08   
1340299        WALLACE DA SILVA SOUZA DA CUNHA 14488599729       08   
1341985                   ARIANE SOUZA REZENDE 13691051774       08   
1349247        JORGE HENRIQUE ALVES DOS SANTOS 12729081780       08   

        motivo_situacao cod_nat_juridica cnae_fiscal tipo_logradouro  \
21503                67             2135     5590699             RUA   
60873                67             2135     4781400             RUA   
71846                67             2135     9602501             RUA   
88541                67             2135     4723700             RUA   
106682               67             2135     4723700             RUA   
...                 ...              ...         ...             ...   
1286786              67             2135     4789099             RUA   
1304084              67             2135     8592999             RUA   
1340299              67             2135     1091102             RUA   
1341985              67             2135     9001901             RUA   
1349247              67             2135     9001901             RUA   

          logradouro numero complemento      bairro       cep  uf  \
21503    SAINT ROMAN    200              COPACABANA  22071060  RJ   
60873    SAINT ROMAN    200              COPACABANA  22071060  RJ   
71846    SAINT ROMAN    200              COPACABANA  22071060  RJ   
88541    SAINT ROMAN    200              COPACABANA  22071060  RJ   
106682   SAINT ROMAN    200              COPACABANA  22071060  RJ   
...              ...    ...         ...         ...       ...  ..   
1286786  SAINT ROMAN    200              COPACABANA  22071060  RJ   
1304084  SAINT ROMAN    200              COPACABANA  22071060  RJ   
1340299  SAINT ROMAN    200              COPACABANA  22071060  RJ   
1341985  SAINT ROMAN    200              COPACABANA  22071060  RJ   
1349247  SAINT ROMAN    200              COPACABANA  22071060  RJ   

              municipio qualif_resp capital_social porte opc_simples opc_mei  \
21503    RIO DE JANEIRO          50        36000.0    01           6   False   
60873    RIO DE JANEIRO          50            1.0    01           6   False   
71846    RIO DE JANEIRO          50            1.0    01           6   False   
88541    RIO DE JANEIRO          50            1.0    01           6   False   
106682   RIO DE JANEIRO          50            1.0    01           6   False   
...                 ...         ...            ...   ...         ...     ...   
1286786  RIO DE JANEIRO          50         1000.0    01           6   False   
1304084  RIO DE JANEIRO          50         1000.0    01           6   False   
1340299  RIO DE JANEIRO          50         5000.0    01           6   False   
1341985  RIO DE JANEIRO          50            1.0    01           6   False   
1349247  RIO DE JANEIRO          50          100.0    01           6   False   

        inexistent

Interessante observar da amostra acima que várias das empresas estão no mesmo endereço, estão baixadas (situação=08 e motivo=67), compartilham a mesma natureza jurídica (2135), e algumas compartilham o mesmo telefone, dentre outras semelhanças.<br>Vamos fazer o merge do consolidado de endereços com a base de empresas:

In [29]:
empresas_df = empresas_df.merge(enderecos_df, on='endereco_processado', how='left')
empresas_df

cnpj matriz_filial  \
0        00000942000170             1   
1        00001102000121             1   
2        00005201000181             1   
3        00005201000262             2   
4        00010501000159             1   
...                 ...           ...   
1559938  98593668000194             1   
1559939  98669997000252             2   
1559940  98670235000195             1   
1559941  98670656000116             1   
1559942  98674195000150             1   

                                              razao_social situacao  \
0                                      NILZA LAVISO ARAUJO       08   
1                          JOSE GERALDO PEREIRA SANTANA ME       08   
2        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO LTDA       08   
3        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO LTDA       08   
4        B S PETRO COM E DIST DE DERIVADOS DE PETROLEO ...       08   
...                                                    ...      ...   
1559938                    AUTO VIACAO VENANCIO AIRES LTDA       02   
1559939          E. R. AMANTINO INDUSTRIA METALURGICA LTDA       02   
1559940                    MAZZAROLLO - AGRO MAQUINAS LTDA       02   
1559941                            RESTAURANTE FARINA LTDA       02   
1559942                           AEROCLUBE DE VERANOPOLIS       02   

        motivo_situacao cod_nat_juridica cnae_fiscal tipo_logradouro  \
0                    67             2135     3299002             RUA   
1                    67             2135     4729699             RUA   
2                    15             2062     4681802             RUA   
3                    15             2062     4681802         AVENIDA   
4                    15             2062     4681801         AVENIDA   
...                 ...              ...         ...             ...   
1559938              00             2062     4922102             RUA   
1559939              00             2062     2451200             RUA   
1559940              00             2062     4789099         AVENIDA   
1559941              00             2062     5611201         ESTRADA   
1559942              00             3999     8599601         ALAMEDA   

                       logradouro numero  complemento          bairro  \
0             PREFEITO S A T PIZA    235               JARDIM PARAIZO   
1                    ANGELO GUIDI    184                       CENTRO   
2                    DOS ANDRADAS    421  CONJUNTO 52  SANTA IFIGENIA   
3        DEPUTADO NARCISO PIERONI    450                    MOSTARDAS   
4                    MONS SECKLER   1494                   VL AMERICA   
...                           ...    ...          ...             ...   
1559938     VOLUNTARIOS DA PATRIA   1889                       CENTRO   
1559939            GETULIO VARGAS    642                    RENOVACAO   
1559940            OSVALDO ARANHA    452                       CENTRO   
1559941           ESTRADA RST 470    691                       CENTRO   
1559942             SANTOS DUMONT    S N                       CENTRO   

              cep  uf            municipio qualif_resp capital_social porte  \
0        16600000  SP              PIRAJUI          50            0.0    01   
1        14180000  SP               PONTAL          50            0.0    01   
2        01208000  SP            SAO PAULO          49            0.0    01   
3        13820000  SP  MONTE ALEGRE DO SUL          49            0.0    01   
4        18540000  SP          PORTO FELIZ          49            0.0    01   
...           ...  ..                  ...         ...            ...   ...   
1559938  95800000  RS       VENANCIO AIRES          49      2000000.0    05   
1559939  95330000  RS          VERANOPOLIS          49            0.0    05   
1559940  95330000  RS          VERANOPOLIS          49      1600000.0    05   
1559941  95330000  RS          VERANOPOLIS          49            0.0    01   
1559942  95330000  RS          VERANOPOLIS          16        

In [30]:
# Conferindo:
empresas_df[empresas_df['endereco_processado'] == 'RUA_SAINT ROMAN_200__COPACABANA_RIO DE JANEIRO_RJ']

cnpj matriz_filial  \
21503    11470156000152             1   
60873    11821417000131             1   
71846    11908988000108             1   
88541    12030632000187             1   
106682   12182277000161             1   
...                 ...           ...   
1286786  21053847000104             1   
1304084  21196378000182             1   
1340299  21502367000183             1   
1341985  21517812000189             1   
1349247  21588278000100             1   

                                              razao_social situacao  \
21503                     ROBERTO VICENTE COSTA DOS SANTOS       08   
60873                 VANESSA RENEE ALVES REIS 44428480253       08   
71846              NIVALDO CAVALCANTE DE MOURA 98736906700       08   
88541                   MAGNO TIAGO DOS SANTOS 01123741743       08   
106682                 ROBSON BRANDAO CASCIANO 07641572723       08   
...                                                    ...      ...   
1286786               LUCIANA LOPES DE ALMEIDA 01490391746       08   
1304084  JOAQUIM GUILHERME BASTOS MARINHO VIANA 1462450...       08   
1340299        WALLACE DA SILVA SOUZA DA CUNHA 14488599729       08   
1341985                   ARIANE SOUZA REZENDE 13691051774       08   
1349247        JORGE HENRIQUE ALVES DOS SANTOS 12729081780       08   

        motivo_situacao cod_nat_juridica cnae_fiscal tipo_logradouro  \
21503                67             2135     5590699             RUA   
60873                67             2135     4781400             RUA   
71846                67             2135     9602501             RUA   
88541                67             2135     4723700             RUA   
106682               67             2135     4723700             RUA   
...                 ...              ...         ...             ...   
1286786              67             2135     4789099             RUA   
1304084              67             2135     8592999             RUA   
1340299              67             2135     1091102             RUA   
1341985              67             2135     9001901             RUA   
1349247              67             2135     9001901             RUA   

          logradouro numero complemento      bairro       cep  uf  \
21503    SAINT ROMAN    200              COPACABANA  22071060  RJ   
60873    SAINT ROMAN    200              COPACABANA  22071060  RJ   
71846    SAINT ROMAN    200              COPACABANA  22071060  RJ   
88541    SAINT ROMAN    200              COPACABANA  22071060  RJ   
106682   SAINT ROMAN    200              COPACABANA  22071060  RJ   
...              ...    ...         ...         ...       ...  ..   
1286786  SAINT ROMAN    200              COPACABANA  22071060  RJ   
1304084  SAINT ROMAN    200              COPACABANA  22071060  RJ   
1340299  SAINT ROMAN    200              COPACABANA  22071060  RJ   
1341985  SAINT ROMAN    200              COPACABANA  22071060  RJ   
1349247  SAINT ROMAN    200              COPACABANA  22071060  RJ   

              municipio qualif_resp capital_social porte opc_simples opc_mei  \
21503    RIO DE JANEIRO          50        36000.0    01           6   False   
60873    RIO DE JANEIRO          50            1.0    01           6   False   
71846    RIO DE JANEIRO          50            1.0    01           6   False   
88541    RIO DE JANEIRO          50            1.0    01           6   False   
106682   RIO DE JANEIRO          50            1.0    01           6   False   
...                 ...         ...            ...   ...         ...     ...   
1286786  RIO DE JANEIRO          50         1000.0    01           6   False   
1304084  RIO DE JANEIRO          50         1000.0    01           6   False   
1340299  RIO DE JANEIRO          50         5000.0    01           6   False   
1341985  RIO DE JANEIRO          50            1.0    01           6   False   
1349247  RIO DE JANEIRO          50          100.0    01           6   False   

        inexistent

Conforme esperado, as 65 empresas.

In [31]:
# Vamos remover colunas não mais necessárias:
cols = ['tipo_logradouro', 'logradouro', 'numero', 'complemento', 'bairro', 'endereco_processado', 'cep']
empresas_df = empresas_df.drop(columns=cols)
# Obs: mantivemos municipio e uf, que são interessantes para análise. Removemos CEP, que não será necessária.

Vamos ver como estamos:

In [32]:
empresas_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1559943 entries, 0 to 1559942
Data columns (total 40 columns):
cnpj                                        1559943 non-null object
matriz_filial                               1559943 non-null object
razao_social                                1559943 non-null object
situacao                                    1559943 non-null object
motivo_situacao                             1559943 non-null object
cod_nat_juridica                            1559943 non-null object
cnae_fiscal                                 1559943 non-null object
uf                                          1559943 non-null object
municipio                                   1559943 non-null object
qualif_resp                                 1559943 non-null object
capital_social                              1559943 non-null object
porte                                       1559943 non-null object
opc_simples                                 1559943 non-null object
opc

Observamos que as colunas que vieram da base de sócios estão com menos 'non-null'. Vamos tratar a seguir.

### Tratamento das demais colunas:

##### Colunas que vieram da base de sócios:

Sobre as colunas que vieram da base de sócios e que ficaram com valores nulos, são empresas para as quais não identificamos os sócios. Vamos preencher os valores com base nas estatísticas da base (média ou moda, conforme cada coluna):

In [33]:
# Calculando a moda das colunas:
modas_df = empresas_df[['qte_socios', 'tempo_de_sociedade', 'auxilio_emergencial_bloqueado', 
                        'socio_solicitou_auxilio_emergencial', 'socio_solicitou_bolsa_familia',
                        'socio_candidato', 'socio_devedor_uniao', 'empresas_inexistentes_socio_participa',
                        'qte_cnpj_socio_participa', 'socio_pessoa_politicamente_exposta',
                        'socio_solicitou_seguro_defeso', 'socio_sancionado']].mode()
modas_df

qte_socios tempo_de_sociedade auxilio_emergencial_bloqueado  \
0          1                  5                             0   

  socio_solicitou_auxilio_emergencial socio_solicitou_bolsa_familia  \
0                                   1                             0   

  socio_candidato socio_devedor_uniao empresas_inexistentes_socio_participa  \
0               0                   0                                     1   

  qte_cnpj_socio_participa socio_pessoa_politicamente_exposta  \
0                        1                                  0   

  socio_solicitou_seguro_defeso socio_sancionado  
0                             0                0

In [34]:
# Preenchendo os nulos:
empresas_df = empresas_df.fillna(value={
    'qte_socios': modas_df.qte_socios[0],
    'tempo_de_sociedade': modas_df.tempo_de_sociedade[0],
    'auxilio_emergencial_bloqueado': modas_df.auxilio_emergencial_bloqueado[0],
    'socio_solicitou_auxilio_emergencial': modas_df.socio_solicitou_auxilio_emergencial[0],
    'socio_solicitou_bolsa_familia': modas_df.socio_solicitou_bolsa_familia[0],
    'socio_candidato': modas_df.socio_candidato[0],
    'socio_devedor_uniao': modas_df.socio_devedor_uniao[0],
    'empresas_inexistentes_socio_participa': modas_df.empresas_inexistentes_socio_participa[0],
    'qte_cnpj_socio_participa': modas_df.qte_cnpj_socio_participa[0],
    'socio_pessoa_politicamente_exposta': modas_df.socio_pessoa_politicamente_exposta[0],
    'socio_recurso_recebido_governo_federal': 0,
    'socio_solicitou_seguro_defeso': modas_df.socio_solicitou_seguro_defeso[0],
    'socio_sancionado': modas_df.socio_sancionado[0]})

# Obs: para 'socio_recurso_recebido_governo_federal', seria usada a média,
# mas não foi possível calculá-la. O notebook sempre travava. Foi assumido o valor 0 para os nulos.

In [35]:
empresas_df.qte_socios = empresas_df.qte_socios.astype(int)
empresas_df.tempo_de_sociedade = empresas_df.tempo_de_sociedade.astype(int)
empresas_df.auxilio_emergencial_bloqueado = empresas_df.auxilio_emergencial_bloqueado.astype(int)
empresas_df.socio_solicitou_auxilio_emergencial = empresas_df.socio_solicitou_auxilio_emergencial.astype(int)
empresas_df.socio_solicitou_bolsa_familia = empresas_df.socio_solicitou_bolsa_familia.astype(int)
empresas_df.socio_candidato = empresas_df.socio_candidato.astype(int)
empresas_df.socio_devedor_uniao = empresas_df.socio_devedor_uniao.astype(int)
empresas_df.empresas_inexistentes_socio_participa = empresas_df.empresas_inexistentes_socio_participa.astype(int)
empresas_df.qte_cnpj_socio_participa = empresas_df.qte_cnpj_socio_participa.astype(int)
empresas_df.socio_pessoa_politicamente_exposta = empresas_df.socio_pessoa_politicamente_exposta.astype(int)
empresas_df.socio_recurso_recebido_governo_federal = empresas_df.socio_recurso_recebido_governo_federal.astype(float)
empresas_df.socio_solicitou_seguro_defeso = empresas_df.socio_solicitou_seguro_defeso.astype(int)
empresas_df.socio_sancionado = empresas_df.socio_sancionado.astype(int)

In [36]:
empresas_df.describe()

qte_socios  tempo_de_sociedade  auxilio_emergencial_bloqueado  \
count  1.559943e+06        1.559943e+06                   1.559943e+06   
mean   1.113958e+00        6.203179e+00                   2.519066e-02   
std    8.809046e-01        3.263609e+00                   1.567039e-01   
min    1.000000e+00        0.000000e+00                   0.000000e+00   
25%    1.000000e+00        5.000000e+00                   0.000000e+00   
50%    1.000000e+00        6.000000e+00                   0.000000e+00   
75%    1.000000e+00        7.000000e+00                   0.000000e+00   
max    3.570000e+02        8.600000e+01                   1.000000e+00   

       socio_solicitou_auxilio_emergencial  socio_solicitou_bolsa_familia  \
count                         1.559943e+06                   1.559943e+06   
mean                          5.454725e-01                   5.951051e-02   
std                           4.979281e-01                   2.365778e-01   
min                           0.000000e+00                   0.000000e+00   
25%                           0.000000e+00                   0.000000e+00   
50%                           1.000000e+00                   0.000000e+00   
75%                           1.000000e+00                   0.000000e+00   
max                           1.000000e+00                   1.000000e+00   

       socio_candidato  socio_devedor_uniao  \
count     1.559943e+06         1.559943e+06   
mean      7.529121e-03         2.747088e-02   
std       8.644327e-02         1.634510e-01   
min       0.000000e+00         0.000000e+00   
25%       0.000000e+00         0.000000e+00   
50%       0.000000e+00         0.000000e+00   
75%       0.000000e+00         0.000000e+00   
max       1.000000e+00         1.000000e+00   

       empresas_inexistentes_socio_participa  qte_cnpj_socio_participa  \
count                           1.559943e+06              1.559943e+06   
mean                            9.054254e-01              2.893540e+00   
std                             3.320688e-01              7.372353e+01   
min                             0.000000e+00              1.000000e+00   
25%                             1.000000e+00              1.000000e+00   
50%                             1.000000e+00              1.000000e+00   
75%                             1.000000e+00              1.000000e+00   
max                             2.300000e+01              1.355900e+04   

       socio_pessoa_politicamente_exposta  \
count                        1.559943e+06   
mean                         1.265431e-03   
std                          3.555039e-02   
min                          0.000000e+00   
25%                          0.000000e+00   
50%                          0.000000e+00   
75%                          0.000000e+00   
max                          1.000000e+00   

       socio_recurso_recebido_governo_federal  socio_solicitou_seguro_defeso  \
count                            1.559943e+06                   1.559943e+06   
mean                             1.922794e+06                   6.275870e-04   
std                              9.308915e+08                   2.504383e-02   
min                             -1.381240e+06                   0.000000e+00   
25%                              0.000000e+00                   0.000000e+00   
50%                              0.000000e+00                   0.000000e+00   
75%                              0.000000e+00                   0.000000e+00   
max                              5.186634e+11                   1.000000e+00   

       socio_sancionado  qte_empresas_com_mesmo_telefone  telefone_preenchido  \
count      1.559943e+06                     1.559943e+06         1.559943e+06   
mean       9.615736e-05                     9.646744e+00         9.866168e-01   
std        9.805518e-03                     1.615946e+02         1.149090e-01   
min        0.000000e+00                     0.000000e+00         0.000000e+00   
25%        0.000000e+00  

Observamos os valores minimos e máximos acima, percebemos que poderíamos ter colocado tipos mais otimizados para as colunas, como np.uint8 ou np.uint32, para economizar memória. Exceto para socio_recurso_recebido_governo_federal, que precisa ser float.
Precisam ser int32: qte_socios, qte_empresas_com_mesmo_telefone e qte_cnpj_socio_participa. Isso será feito posteriormente.

### Binarização de colunas:

##### Coluna opc_mei:<br>
Pelo dicionário de dados, os valores são: S-SIM; N-NÃO; OUTROS (BRANCO, ETC) Podemos reduzir para uma variável booleana 1-OPÇÃO PELO MEI; 0-NÃO OPÇÃO PELO MEI

In [37]:
# Antes da binarizarção:
empresas_df.opc_mei.value_counts()

False    1533044
True       26899
Name: opc_mei, dtype: int64

In [38]:
# Binarização:
empresas_df.opc_mei = np.where(empresas_df.opc_mei == 'True', 1, 0).astype(np.uint8)
empresas_df.opc_mei.value_counts()

0    1533044
1      26899
Name: opc_mei, dtype: int64

##### Coluna 'inexistente':

In [39]:
# Antes da binarizarção:
empresas_df.inexistente.value_counts()

True     1375877
False     184066
Name: inexistente, dtype: int64

In [40]:
# Binarização:
empresas_df.inexistente = np.where(empresas_df.inexistente == 'True', 1, 0).astype(np.uint8)
empresas_df.inexistente.value_counts()

1    1375877
0     184066
Name: inexistente, dtype: int64

##### Coluna 'telefone_provavelmente_invalido':

In [41]:
# Antes da binarizarção:
empresas_df.telefone_provavelmente_invalido.value_counts()

False    1551495
True        8448
Name: telefone_provavelmente_invalido, dtype: int64

In [42]:
# Binarização:
empresas_df.telefone_provavelmente_invalido = np.where(empresas_df.telefone_provavelmente_invalido == 'True', 1, 0).astype(np.uint8)
empresas_df.telefone_provavelmente_invalido.value_counts()

0    1551495
1       8448
Name: telefone_provavelmente_invalido, dtype: int64

Fim da binarização.

### Limpeza e tipificação das demais colunas:

In [43]:
# Coluna 'situacao'
print(empresas_df.situacao.value_counts())
empresas_df.situacao = empresas_df.situacao.astype(np.uint8)
print(empresas_df.situacao.value_counts())

08    1385738
02     167207
04       5939
01        720
03        339
Name: situacao, dtype: int64
8    1385738
2     167207
4       5939
1        720
3        339
Name: situacao, dtype: int64


In [44]:
# Coluna 'motivo_situacao'
print(empresas_df.motivo_situacao.value_counts())
empresas_df.motivo_situacao = empresas_df.motivo_situacao.astype('category')

67    1372057
00     167207
01       8831
63       5829
15       3226
02        810
16        690
71        308
73        306
54        175
18        159
21        123
64         72
36         25
74         21
70         21
10         17
37         14
04         11
17          8
05          8
34          6
72          5
07          3
11          3
06          2
14          2
09          1
24          1
61          1
80          1
Name: motivo_situacao, dtype: int64


In [45]:
# Coluna cod_nat_juridica
print(empresas_df.cod_nat_juridica.value_counts())
empresas_df.cod_nat_juridica = empresas_df.cod_nat_juridica.astype('category')

2135    1430198
2062      90698
2305      21979
2054       5835
3999       2304
         ...   
2089          1
3255          1
1198          1
2097          1
8885          1
Name: cod_nat_juridica, Length: 65, dtype: int64


In [46]:
# Coluna cnae_fiscal
print(empresas_df.cnae_fiscal.value_counts())
empresas_df.cnae_fiscal = empresas_df.cnae_fiscal.astype('category')

4781400    151887
9602501     93445
4399103     54139
5611203     46133
7911200     36747
            ...  
0321303         1
0729401         1
2531402         1
0159899         1
6499902         1
Name: cnae_fiscal, Length: 1232, dtype: int64


In [47]:
# Coluna uf
print(empresas_df.uf.value_counts())
empresas_df.uf = empresas_df.uf.astype('category')

SP    411029
RJ    231155
MG    140327
BA     94877
RS     83490
PR     70916
GO     55759
PE     52122
PA     48356
CE     47399
SC     44965
ES     37071
DF     30869
MT     28375
MA     25288
MS     22218
AM     21750
RN     19343
PB     18757
AL     17759
PI     12091
RO     11646
TO     10918
SE      9671
AP      5163
AC      4980
RR      3623
EX        26
Name: uf, dtype: int64


In [48]:
# Coluna municipio
print(empresas_df.municipio.value_counts())
empresas_df.municipio = empresas_df.municipio.astype('category')

SAO PAULO                     143300
RIO DE JANEIRO                100113
SALVADOR                       36393
BRASILIA                       30869
BELO HORIZONTE                 30310
                               ...  
LASTRO                             1
SANTA FILOMENA DO MARANHAO         1
OLIVEDOS                           1
ENCANTO                            1
SUCUPIRA DO RIACHAO                1
Name: municipio, Length: 5280, dtype: int64


In [49]:
# Coluna qualif_resp
print(empresas_df.qualif_resp.value_counts())
empresas_df.qualif_resp = empresas_df.qualif_resp.astype('category')

50    1430143
49      84712
65      21818
05       9932
16       7887
10       3956
19        475
64        463
59        227
12         62
17         59
13         57
28         42
11         35
34         26
42          9
32          8
40          5
00          5
39          5
31          5
23          3
36          2
51          2
43          2
54          2
24          1
Name: qualif_resp, dtype: int64


In [50]:
# Coluna opc_simples
print(empresas_df.opc_simples.value_counts())
empresas_df.opc_simples = empresas_df.opc_simples.astype('category')

6    1399627
5     107901
0      52415
Name: opc_simples, dtype: int64


In [51]:
# Coluna porte
print(empresas_df.porte.value_counts())
empresas_df.porte = empresas_df.porte.astype('category')

01    1493736
05      43857
03      22344
00          6
Name: porte, dtype: int64


In [52]:
# Coluna matriz_filial
print(empresas_df.matriz_filial.value_counts())
empresas_df.matriz_filial = empresas_df.matriz_filial.astype('category')

1    1547369
2      12574
Name: matriz_filial, dtype: int64


In [53]:
# Coluna email_processado
print(empresas_df.email_processado.value_counts())
empresas_df.email_processado = empresas_df.email_processado.astype('category')

NI                             611623
HOTMAIL.COM                    447041
GMAIL.COM                      188999
YAHOO.COM.BR                    71222
BOL.COM.BR                      29532
                                ...  
ASATNET.COM.BR                      1
AROXA.COM.BR                        1
CONCEITOASSESSORIASA.COM.BR         1
LIBERTYTURISMO.COM                  1
BRAVIETOUR.COM.BR                   1
Name: email_processado, Length: 53000, dtype: int64


In [54]:
# As colunas abaixo também podem ser consideradas como uint8 para redução de memória (são basicamente binárias):
cols = ['empresa_devedora_uniao', 'empresa_venceu_licitacao', 'empresa_participou_licitacao', 'empresa_sancionada',
'auxilio_emergencial_bloqueado', 'socio_solicitou_auxilio_emergencial', 'socio_solicitou_bolsa_familia', 
'socio_candidato', 'socio_devedor_uniao']
for col in cols:
    empresas_df[col] = empresas_df[col].astype(np.uint8)
# Já as colunas abaixo precisam ser maiores, por poderem conter valores maiores (quantidade de empresas, sócios e atendimentos):
cols = ['qte_socios', 'qte_empresas_com_mesmo_telefone', 'qte_cnpj_socio_participa', 'Atendimentos_Procon']
for col in cols:
    empresas_df[col] = empresas_df[col].astype(np.uint32)

In [55]:
# Coluna empresas_df.anos_de_atividade_empresa
print(empresas_df.anos_de_atividade_empresa.max())
print(empresas_df.anos_de_atividade_empresa.min())
empresas_df.anos_de_atividade_empresa = empresas_df.anos_de_atividade_empresa.astype(np.uint8)

96
0


Fim da limpeza e tipificação de colunas.

##### Análise da coluna capital_social:

In [56]:
# Coluna capital_social
empresas_df.capital_social = empresas_df.capital_social.astype(float)

In [57]:
empresas_df['capital_social'].describe()

count    1.559943e+06
mean     2.704682e+06
std      2.715899e+08
min      0.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      5.000000e+03
max      2.054320e+11
Name: capital_social, dtype: float64

Vemos acima que existem valores muito distorcidos. Vamos identificar outliers:

In [58]:
# Identificação de limites para remoção de outliers do capital social:
Q3 = empresas_df['capital_social'].quantile(0.75)
Q1 = empresas_df['capital_social'].quantile(0.2)
IQR = Q3 - Q1
quartil_superior = Q3 + 1.5 * IQR
quartil_inferior = Q1 - 1.5 * IQR
print(quartil_superior, quartil_inferior)

12498.5 -7497.5


Esses valores não estão refletindo a realidade de capital social das empresas. Vamos optar por remover a coluna de capital social.

In [59]:
empresas_df = empresas_df.drop(columns=['capital_social'])

In [60]:
empresas_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1559943 entries, 0 to 1559942
Data columns (total 39 columns):
cnpj                                        1559943 non-null object
matriz_filial                               1559943 non-null category
razao_social                                1559943 non-null object
situacao                                    1559943 non-null uint8
motivo_situacao                             1559943 non-null category
cod_nat_juridica                            1559943 non-null category
cnae_fiscal                                 1559943 non-null category
uf                                          1559943 non-null category
municipio                                   1559943 non-null category
qualif_resp                                 1559943 non-null category
porte                                       1559943 non-null category
opc_simples                                 1559943 non-null category
opc_mei                                     1559943 no

Observa-se que na primeira importação a memória ocupada era de: 583.2+ MB<br>
Agora, estamos com 177.2MB.

In [61]:
empresas_df

cnpj matriz_filial  \
0        00000942000170             1   
1        00001102000121             1   
2        00005201000181             1   
3        00005201000262             2   
4        00010501000159             1   
...                 ...           ...   
1559938  98593668000194             1   
1559939  98669997000252             2   
1559940  98670235000195             1   
1559941  98670656000116             1   
1559942  98674195000150             1   

                                              razao_social  situacao  \
0                                      NILZA LAVISO ARAUJO         8   
1                          JOSE GERALDO PEREIRA SANTANA ME         8   
2        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO LTDA         8   
3        PETRO-ALFA COMERCIO DE DERIVADOS DE PETROLEO LTDA         8   
4        B S PETRO COM E DIST DE DERIVADOS DE PETROLEO ...         8   
...                                                    ...       ...   
1559938                    AUTO VIACAO VENANCIO AIRES LTDA         2   
1559939          E. R. AMANTINO INDUSTRIA METALURGICA LTDA         2   
1559940                    MAZZAROLLO - AGRO MAQUINAS LTDA         2   
1559941                            RESTAURANTE FARINA LTDA         2   
1559942                           AEROCLUBE DE VERANOPOLIS         2   

        motivo_situacao cod_nat_juridica cnae_fiscal  uf            municipio  \
0                    67             2135     3299002  SP              PIRAJUI   
1                    67             2135     4729699  SP               PONTAL   
2                    15             2062     4681802  SP            SAO PAULO   
3                    15             2062     4681802  SP  MONTE ALEGRE DO SUL   
4                    15             2062     4681801  SP          PORTO FELIZ   
...                 ...              ...         ...  ..                  ...   
1559938              00             2062     4922102  RS       VENANCIO AIRES   
1559939              00             2062     2451200  RS          VERANOPOLIS   
1559940              00             2062     4789099  RS          VERANOPOLIS   
1559941              00             2062     5611201  RS          VERANOPOLIS   
1559942              00             3999     8599601  RS          VERANOPOLIS   

        qualif_resp porte opc_simples  opc_mei  inexistente  \
0                50    01           6        0            1   
1                50    01           6        0            1   
2                49    01           0        0            1   
3                49    01           0        0            1   
4                49    01           0        0            1   
...             ...   ...         ...      ...          ...   
1559938          49    05           0        0            0   
1559939          49    05           0        0            0   
1559940          49    05           0        0            0   
1559941          49    01           5        0            0   
1559942          16    05           0        0            0   

         anos_de_atividade_empresa  telefone_provavelmente_invalido  \
0                               24                                0   
1                               24                                0   
2                               24                                0   
3                               17                                0   
4                               24                                0   
...                            ...                              ...   
1559938                         53                                0   
1559939                         39                                0   
1559940                         55                                0   
1559941                         55                                0   
1559942                         54                                0   

         Atendimentos_Procon  empresa_devedora_uniao  \
0                     

In [62]:
# Salvando resultado
empresas_df.to_csv(r'C:\DATASETS\Empresas Rotuladas\empresas_rotuladas-processado.csv', index = False)

Fim

Para quando for necessário reimportar a base:

In [ ]:
# REIMPORTAÇÃO DA BASE DE EMPRESAS - Estamos informando os tipos já conhecidos previamente.
colstype = {'cnpj': str, 'matriz_filial': category, 'razao_social': str, 'situacao': np.uint8, 'motivo_situacao': category,
'cod_nat_juridica': category, 'cnae_fiscal': category, 'uf': category, 'municipio': category, 'qualif_resp': category,
'porte': category, 'opc_simples': category, 'opc_mei': np.uint8, 'inexistente': np.uint8, 'anos_de_atividade_empresa': np.uint8,
'telefone_provavelmente_invalido': np.uint8, 'Atendimentos_Procon': np.uint32, 'empresa_devedora_uniao': np.uint8,
'empresa_venceu_licitacao': np.uint8, 'empresa_participou_licitacao': np.uint8, 'empresa_recurso_recebido_governo_federal': float,
'empresa_sancionada': np.uint8, 'qte_socios': np.uint32, 'tempo_de_sociedade': np.int32, 'auxilio_emergencial_bloqueado': np.int8, 
'socio_solicitou_auxilio_emergencial': np.int8, 'socio_solicitou_bolsa_familia': np.int8, 'socio_candidato': np.int8, 
'socio_devedor_uniao': np.int8, 'empresas_inexistentes_socio_participa': np.int32, 'qte_cnpj_socio_participa': np.int32,
'socio_pessoa_politicamente_exposta': np.int8, 'socio_recurso_recebido_governo_federal': float, 
'socio_solicitou_seguro_defeso': np.int8, 'socio_sancionado': np.int8, 'email_processado': category, 
'qte_empresas_com_mesmo_telefone': : np.int32, 'telefone_preenchido': np.int8, 'qte_empresas_com_mesmo_endereco': np.int32}
empresas_df = pd.read_csv(r'C:\DATASETS\Empresas Rotuladas\empresas_rotuladas-processado.csv', dtype=colstype)